oximeter/oximeter/src/test_util.rs

```diff
@@ -104,3 +114,169 @@ pub fn generate_test_samples(
+///
+/// Similar to the [`expectorate`][1] crate, users can automatically overwrite
+/// the file by setting the environment variable `EXPECTORATE=overwrite` when
+/// running `cargo test`.
+///
+/// [1]: <https://docs.rs/expectorate/latest/expectorate/>
+pub fn check_metric_schema(
+    path: impl AsRef<Path>,
+    metric: impl Metric,
+) -> anyhow::Result<()> {
+    let actual = serde_json::json!({
+        "name": metric.name(),
+        "version": metric.version(),
+        "field_names": metric
+            .field_names()
+            .into_iter()
+            .map(ToString::to_string)
+            .collect::<Vec<_>>(),
+        "field_types": metric
+            .field_types()
+            .into_iter()
+            .map(|f| format!("{f:?}"))
+            .collect::<Vec<_>>(),
+        "datum_type": metric.datum_type(),
+    });
+    check_contents(path, actual)
+}
+
+fn check_contents(
+    path: impl AsRef<Path>,
+    actual: serde_json::Value,
+) -> anyhow::Result<()> {
+    let actual_str = serde_json::to_string_pretty(&actual)
+        .context("failed to serialize JSON")?;
+
+    // Try to read the file.
+    //
+    // If this fails, either to read at all or parse as JSON, overrwrite it and
+    // succeed if the `EXPECTORATE=overwrite` environment variable is set.
+    let is_overwrite =
+        std::env::var("EXPECTORATE").map(|v| v == "overwrite").unwrap_or(false);
+    let contents = match std::fs::read_to_string(path.as_ref()) {
+        Ok(s) => s,
+        Err(_) => {
+            if is_overwrite {
+                std::fs::write(path, &actual_str)
+                    .context("failed to overwrite schema file")?;
+                return Ok(());
+            }
+            anyhow::bail!(
+                "Failed to read schema file. Set \
+                EXPECTORATE=overwrite if you'd like \
+                to overwrite the file anyway"
+            );
+        }
+    };
+    let expected: serde_json::Value = match serde_json::from_str(&contents) {
+        Ok(v) => v,
+        Err(_) => {
+            if is_overwrite {
+                std::fs::write(path, &actual_str)
+                    .context("failed to overwrite schema file")?;
+                return Ok(());
+            }
+            anyhow::bail!(
+                "Failed to read schema file. Set \
+                EXPECTORATE=overwrite if you'd like \
+                to overwrite the file anyway"
+            );
+        }
+    };
+    if expected == actual {
+        return Ok(());
+    }
+
+    // If the contents have changed, we ensure that the version is also greater.
+    let (
+        serde_json::Value::Number(expected_version),
+        serde_json::Value::Number(actual_version)) = (&expected["version"], &actual["version"]) else {
+        anyhow::bail!("expected versions to be numbers");
+    };
+    if expected_version.as_i64().unwrap() >= actual_version.as_i64().unwrap()
+        && !is_overwrite
+    {
```

Comment:
If "expected >= actual && !overwrite", we throw an error. That makes sense to me.

But if "expected < actual", why don't we also throw an error? Why is that case allowed?

Reply:
The idea was that you've properly increased the version along with a change to the schema. This is part of why I didn't just use `expectorate` here: I want to check that the contents are the same, or if they're _not_, the version number is larger. Part of my worry was that it seems likely that folks just kinda put `EXPECTORATE=overwrite` whenever they see a mismatch in this kind of situation. I was hoping it'd be better for folks to see an error that they need to bump the version, and then do that, at which point the update will happen automatically.

Reply:
But it seems like setting `EXPECTORATE=overwrite` obliterates the stored thing anyway -- so ironically, if someone skims the error, sees a diff, and sets that environment variable, they'll toss out the old stored JSON version and just use whatever value came from the in-memory struct regardless.

This could arguably even happen accidentally, with someone changing an unrelated openapi definition, but running `EXPECTORATE=ovewrite cargo test` and trying to bump everything.

I know we're trying to go for a better user experience here, but I think the guidance is on "what you should change" is more subtle than "compare what happened to exist in this file before". I think this is complicated by the file not always representing "what is stored on upstream" -- If I update `http-service.json` to use `version=2`, and then update it again before merging, showing the diff with this file isn't quite right, I actually care about the diff from this file on `main`.

Reply:
FWIW, I am also interested in @ahl 's thoughts here, since our current policy everywhere else is:

- Store the schema explicitly
- Override the schema entirely
- Rely on humans to check the diff during code review, and confirm that these migrations are okay

Reply:
I agree with all that, and I definitely waffled when implementing it. So the goal is to make sure that folks don't change the schema willy-nilly. If we do, we _really_ want to make sure that (1) they are aware of that and (2) they've definitely bumped the version.

If we're relying on humans checking the diff, then straight-up `expectorate` seems like the easiest / simplest approach. Adam's thoughts would definitely be helpful here.

> If I update http-service.json to use version=2, and then update it again before merging, showing the diff with this file isn't quite right, I actually care about the diff from this file on main.

So for the multiple updates issue, we could explicitly compare with the checked in contents. I'm not sure if that's better, or just more magical though.

Reply:
Yeah, I was on the fence about whether or not to suggest that, but it does seem like it encapsulates we care about:

1. Does your "actual struct" match your "schema file"? (e.g., do you acknowledge you are changing the schema, in some way?)
2. Does your schema change (diff with main) update the version, if any diff exists? (e.g., if you're changing the schema, are you doing it the right way?)

Reply:
I talked with Adam about some of this. He threw out a possibly crazy idea: suppose we don't have human-friendly versions for these, but we use a digest of the timeseries schema itself?

Some drawbacks

- Lose human-friendly version numbers. Kind of annoying to look at `target@111683034273293` instead of `target@2`
- The ordering of the versions has to be done by the time they're created (stored in the DB), not the versions
- There is no default. Since we're _adding_ versions now, we can't really generate any meaningful default for those schema which lack versions. This is the big one, IMO. I think we'd need to just blow away data prior to this, or make 0 special and hope no schema actually hash to that.
- We need to pick a stable hash, something like a SHA sum, since the hash implementation in the Rust standard library could change.

Some benefits

- It's not really possible to change the schema and _not_ change the version. We may still rely on a file for developers to ensure that their schema only changes when they want it to, but it's not possible to hit schema mismatches in the database when trying to insert new data.
- Developers don't have to worry about the version for the most part. It changes when it needs to.

This _seems_ like it might make querying harder, but I don't believe it actually does. We _can_ still support querying a specific version of the timeseries, a client just needs to use a bigger number to do that. But supporting a querying against the timeseries _agnostic to the version_ is no different. We probably still want to go look at all the versions for the timeseries to resolve those that are consistent with the query (if any).

Reply:
I talked with Adam about some of this. He threw out a possibly crazy idea: suppose we don't have human-friendly versions for these, but we use a digest of the timeseries schema itself?

Some drawbacks

- Lose human-friendly version numbers. Kind of annoying to look at `target@111683034273293` instead of `target@2`
- The ordering of the versions has to be done by the time they're created (stored in the DB), not the versions
- There is no default. Since we're _adding_ versions now, we can't really generate any meaningful default for those schema which lack versions. This is the big one, IMO. I think we'd need to just blow away data prior to this, or make 0 special and hope no schema actually hash to that.
- We need to pick a stable hash, something like a SHA sum, since the hash implementation in the Rust standard library could change.

Some benefits

- It's not really possible to change the schema and _not_ change the version. We may still rely on a file for developers to ensure that their schema only changes when they want it to, but it's not possible to hit schema mismatches in the database when trying to insert new data.
- Developers don't have to worry about the version for the most part. It changes when it needs to.

This _seems_ like it might make querying harder, but I don't believe it actually does. We _can_ still support querying a specific version of the timeseries, a client just needs to use a bigger number to do that. But supporting a querying against the timeseries _agnostic to the version_ is no different. We probably still want to go look at all the versions for the timeseries to resolve those that are consistent with the query (if any).